File: app/src/main/java/line/tree/Tree.java
package line.tree;

/**
 * Created by ying.fu on 2018/7/16.
 */

public class Tree {
    private TreeNode mRoot;

    public TreeNode getRoot() {
        return mRoot;
    }

    public void setRoot(TreeNode root) {
        mRoot = root;
    }

    public int getLevel() {
        if (mRoot == null)
            return 0;
        else
            return getLevel(1, mRoot);
    }

    public boolean isLeaf(TreeNode node) {
        if (node != null && node.getLeftNode() == null && node.getRightNode() == null)
            return true;
        return false;
    }

    public int getLevel(int level, TreeNode node) {
        if (isLeaf(node)) {
            return level;
        } else {
            level = level + 1;
            level = Math.max(getLevel(level, node.getLeftNode()), getLevel(level, node.getRightNode()));
            return level;
        }
    }

    String s = "";

    public String preTraversal() {
        s = "";
        if (mRoot == null)
            return s;
        else
            s = preTraversal(mRoot);
        return s;
    }


    private String preTraversal(TreeNode node) {
        s = s + node.getValue() + ", ";
        if (!isLeaf(node)) {
            preTraversal(node.getLeftNode());
            preTraversal(node.getRightNode());
        }
        return s;
    }

    public String inTraversal() {
        s = "";
        if (mRoot == null)
            return s;
        else {
            s = inTraversal(mRoot);
        }
        return s;
    }

    private String inTraversal(TreeNode node) {
        if (!isLeaf(node)) {
            inTraversal(node.getLeftNode());
            s = s + node.getValue() + ", ";
            inTraversal(node.getRightNode());
        } else {
            s = s + node.getValue() + ", ";
        }
        return s;
    }

    public String postTraversal() {
        s = "";
        if (mRoot == null)
            return s;
        else
            s = postTraversal(mRoot);
        return s;
    }

    private String postTraversal( TreeNode node) {
        if (!isLeaf(node)) {
            postTraversal(node.getLeftNode());
            postTraversal(node.getRightNode());
            s = s + node.getValue() + ", ";
        } else {
            s = s + node.getValue() + ", ";
        }
        return s;
    }


}

File: javalib/src/main/java/com/yin/javalib/tree/Tree.kt
package com.yin.javalib.tree

/**
 * Created by ying.fu on 2018/7/16.
 */
class Tree {
    var root: TreeNode? = null

    val level: Int
        get() = if (this.root == null) 0 else getLevel(1, this.root!!)

    private fun isLeaf(node: TreeNode?): Boolean {
        return node != null && node.leftNode == null && node.rightNode == null
    }

    private fun getLevel(level: Int, node: TreeNode?): Int {
        var level = level
        return if (isLeaf(node)) {
            level
        } else {
            level += 1
            level = getLevel(level, node?.leftNode).coerceAtLeast(getLevel(level, node?.rightNode))
            level
        }
    }

    private var s = ""
    fun preTraversal(): String {
        s = ""
        s = if (this.root == null) return s else preTraversal(this.root!!)
        return s
    }

    private fun preTraversal(node: TreeNode): String {
        s = s + node.value + ", "
        if (!isLeaf(node)) {
            preTraversal(node.leftNode!!)
            preTraversal(node.rightNode!!)
        }
        return s
    }

    fun inTraversal(): String {
        s = ""
        s = if (this.root == null) return s else {
            inTraversal(this.root!!)
        }
        return s
    }

    private fun inTraversal(node: TreeNode): String {
        if (!isLeaf(node)) {
            inTraversal(node.leftNode!!)
            s = s + node.value + ", "
            inTraversal(node.rightNode!!)
        } else {
            s = s + node.value + ", "
        }
        return s
    }

    fun postTraversal(): String {
        s = ""
        s = if (this.root == null) return s else postTraversal(this.root!!)
        return s
    }

    private fun postTraversal(node: TreeNode): String {
        s = if (!isLeaf(node)) {
            postTraversal(node.leftNode!!)
            postTraversal(node.rightNode!!)
            s + node.value + ", "
        } else {
            s + node.value + ", "
        }
        return s
    }
}